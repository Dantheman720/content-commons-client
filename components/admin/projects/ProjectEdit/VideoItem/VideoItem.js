/**
 *
 * VideoItem
 *
 */
import React, { Fragment } from 'react';
import { bool, func, string } from 'prop-types';
import {
  Icon, Loader, Progress
} from 'semantic-ui-react';

import './VideoItem.scss';

import { projects } from 'components/admin/projects/ProjectEdit/mockData';

/* eslint-disable react/prefer-stateless-function */
class VideoItem extends React.PureComponent {
  constructor( props ) {
    super( props );

    /**
     * @todo simulate upload for dev purposes;
     * replace for production
     * 1MB = 1,048,576 Bytes
     */
    this.MEGABYTE = 1048576;
    this.MIN_INTERVAL = 500;
    this.MAX_INTERVAL = 1500;
    this.MIN_MB_SEC = 400;
    this.MAX_MB_SEC = 500;

    this.state = {
      video: this.getVideoItem(),
      bytesUploaded: 0,
      nIntervId: null,
      isUploading: false
    };
  }

  componentDidMount = () => {
    /**
     * @todo simulate upload for dev purposes;
     * replace for production
     * min, max interval in milliseconds
     */
    const interval = this.getRandomInt( this.MIN_INTERVAL, this.MAX_INTERVAL );
    const nIntervId = setInterval( this.uploadItem, interval );
    this.setState( { nIntervId } );
  }

  componentWillUnmount = () => {
    clearInterval( this.state.nIntervId );
  }

  /**
   * @todo Replace after GraphQL & Apollo
   */
  getVideoItem = () => {
    const { itemId } = this.props;
    const { videos } = projects[0];
    return videos.find( video => video.id === itemId );
  };

  /**
   * @todo simulate upload for dev purposes;
   * replace for production
   */
  getRandomInt = ( min, max ) => {
    const minNum = Math.ceil( min );
    const maxNum = Math.floor( max );
    return (
      Math.floor( Math.random() * ( ( maxNum - minNum ) + 1 ) ) + minNum
    );
  }

  getRemainingUnits = ( totalUnits, unitsUploaded ) => (
    totalUnits - unitsUploaded
  )

  incrementUpload = ( unit, min, max ) => (
    this[unit] * this.getRandomInt( min, max )
  )

  endUpload = intervId => {
    clearInterval( intervId );
  };

  uploadItem = () => {
    /**
     * @todo simulate upload for dev purposes;
     * replace for production
     * min, max increment in megabytes
     */
    this.setState( nextState => {
      const { filesize } = this.state.video.source[0].size;
      const { bytesUploaded } = nextState;

      const remainingBytes = this.getRemainingUnits( filesize, bytesUploaded );
      let increment = this.incrementUpload( 'MEGABYTE', this.MIN_MB_SEC, this.MAX_MB_SEC );

      if ( remainingBytes < increment ) {
        increment = remainingBytes;
      }

      // continue uploading
      if ( bytesUploaded < filesize ) {
        return {
          bytesUploaded: this.state.bytesUploaded + increment,
          isUploading: true
        };
      }

      // stop uploading
      this.endUpload( nextState.nIntervId );
      return { isUploading: false };
    } );
  }

  render() {
    const {
      onClick,
      displayItemInModal
    } = this.props;

    const { video } = this.state;

    if ( !video || video.loading ) {
      return (
        <Loader active inline="centered">
          <p style={ { fontSize: '0.75em' } }>
            Preparing file for upload...
          </p>
        </Loader>
      );
    }

    const {
      title,
      language,
      thumbnail,
      alt,
      fileName,
      source,
      error,
      uploadStatus
    } = video;

    const { filesize } = source[0].size;
    const { bytesUploaded, isUploading } = this.state;

    const itemStyle = {
      cursor: isUploading ? 'not-allowed' : 'pointer'
    };
    if ( !displayItemInModal ) itemStyle.cursor = 'default';

    const uploadingClass = isUploading ? ' isUploading' : '';
    const Wrapper = !isUploading && displayItemInModal ? 'button' : 'span';
    const wrapperClass = displayItemInModal ? 'modal-trigger' : 'wrapper';

    if ( error || uploadStatus.error ) {
      return (
        <li className="item video error" style={ { textAlign: 'center' } }>
          <Icon
            color="red"
            name="exclamation triangle"
            size="large"
          />
          <p style={ { fontSize: '0.75em' } }>
            { `${error ? 'A loading' : 'An uploading'} error occurred for this item.` }
          </p>
        </li>
      );
    }

    return (
      <li className="item video">
        <Wrapper
          className={ wrapperClass }
          onClick={
            !isUploading && displayItemInModal ? onClick : null
          }
          style={ itemStyle }
        >
          <div className={ `thumbnail${uploadingClass}` }>
            { thumbnail
              && (
                <Fragment>
                  <img src={ thumbnail } alt={ alt } />
                  <p className="file-name">{ fileName }</p>
                </Fragment>
              ) }
            { isUploading
              && (
                <div className="loading-animation">
                  <Loader active inline="centered" />
                </div>
              ) }
          </div>
          { isUploading
            && (
              <Progress
                value={ bytesUploaded }
                total={ filesize }
                color="blue"
                size="small"
                active
                progress
                precision={ 0 }
              >
                <p>Upload in progress</p>
              </Progress>
            ) }
          <h3 className={ `item-heading ${language.text_direction}` }>{ title }</h3>
          <p className="item-lang">{ language.display_name }</p>
        </Wrapper>
      </li>
    );
  }
}

VideoItem.propTypes = {
  // video: object,
  displayItemInModal: bool,
  onClick: func,
  // videoID: string.isRequired,
  itemId: string.isRequired
};

export default VideoItem;
