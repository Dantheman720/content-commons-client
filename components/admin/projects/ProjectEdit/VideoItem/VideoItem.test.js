import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { Icon, Loader } from 'semantic-ui-react';
import VideoItem from './VideoItem';

const props = {
  displayItemInModal: true,
  onClick: jest.fn(),
  itemId: '3728'
};

const Component = <VideoItem { ...props } />;

describe( '<VideoItem />', () => {
  it( 'renders without crashing', () => {
    const wrapper = shallow( Component );
    expect( wrapper.exists() ).toEqual( true );
    expect( toJSON( wrapper ) ).toMatchSnapshot();
  } );

  it( 'renders `Progress` component if `isUploading`', () => {
    const wrapper = shallow( Component );
    const progress = () => wrapper.find( 'Progress' );
    const isUploading = () => wrapper.state( 'isUploading' );

    expect( isUploading() ).toEqual( false );
    expect( progress().exists() ).toEqual( isUploading() );

    wrapper.setState( { isUploading: true } );
    expect( progress().exists() ).toEqual( isUploading() );
    expect( progress().prop( 'total' ) )
      .toEqual( wrapper.state( 'video' ).source[0].size.filesize );
    expect( toJSON( wrapper ) ).toMatchSnapshot();
  } );

  it( 'renders a `span` component with correct props if `isUploading`', () => {
    const wrapper = shallow( Component );
    wrapper.setState( { isUploading: true } );
    const span = wrapper.childAt( 0 );

    expect( span.exists() ).toEqual( true );
    expect( span.hasClass( 'modal-trigger' ) ).toEqual( true );
    expect( span.prop( 'onClick' ) ).toEqual( null );
    expect( span.prop( 'style' ) )
      .toEqual( { cursor: 'not-allowed' } );
  } );

  it( 'renders a `button` with correct props if `displayItemInModal`', () => {
    const wrapper = shallow( Component );
    const button = wrapper.childAt( 0 );

    expect( button.exists() ).toEqual( true );
    expect( button.prop( 'onClick' )._isMockFunction )
      .toEqual( true );
    expect( button.hasClass( 'modal-trigger' ) ).toEqual( true );
    expect( button.prop( 'style' ) )
      .toEqual( { cursor: 'pointer' } );
  } );

  it( 'clicking the `button` calls `onClick`', () => {
    const wrapper = shallow( Component );
    const button = wrapper.childAt( 0 );

    button.simulate( 'click' );
    expect( button.exists() ).toEqual( true );
    expect( button.prop( 'onClick' ) ).toHaveBeenCalled();
  } );

  it( 'renders a `span` with correct props if `displayItemInModal` is `false`', () => {
    const wrapper = shallow( Component );
    wrapper.setProps( { displayItemInModal: false } );
    const span = wrapper.childAt( 0 );

    expect( span.exists() ).toEqual( true );
    expect( span.hasClass( 'wrapper' ) ).toEqual( true );
    expect( span.prop( 'onClick' ) ).toEqual( null );
    expect( span.prop( 'style' ) )
      .toEqual( { cursor: 'default' } );
    expect( toJSON( wrapper ) ).toMatchSnapshot();
  } );

  it( 'renders null if `video` state is empty', () => {
    const wrapper = shallow( Component );

    wrapper.setState( { video: {} } );
    expect( wrapper.state( 'video' ) ).toEqual( {} );
    expect( wrapper.html() ).toEqual( null );
  } );

  it( 'renders null if there is no `video` state', () => {
    const wrapper = shallow( Component );

    wrapper.setState( { video: null } );
    expect( wrapper.state( 'video' ) ).toEqual( null );
    expect( wrapper.html() ).toEqual( null );
  } );

  it( 'renders an uploading message and `Loader` if `video` is loading', () => {
    const wrapper = shallow( Component );

    wrapper.setState( {
      video: {
        id: '3728',
        title: 'Made in America',
        uploadStatus: {
          error: false,
          success: false
        },
        loading: true
      }
    } );

    expect( wrapper.state( 'video' ).loading )
      .toEqual( true );
    expect( wrapper.contains(
      <Loader active inline="centered">
        <p style={ { fontSize: '0.75em' } }>
          Preparing file for upload...
        </p>
      </Loader>
    ) ).toEqual( true );
  } );

  it( 'renders error icon and message if `video` has an error `uploadStatus`', () => {
    const wrapper = shallow( Component );

    expect( wrapper.find( 'li.video.error' ).exists() )
      .toEqual( false );

    wrapper.setState( {
      video: {
        id: '3728',
        title: 'Made in America',
        uploadStatus: {
          error: true,
          success: false
        },
        source: [
          {
            burnedInCaptions: 'false',
            downloadUrl: 'https://video-download-url.com',
            duration: '9:16',
            filetype: '',
            md5: '',
            fileName: 'madeinamerica_english.mp4',
            size: {
              bitrate: 9832917,
              filesize: 662595174,
              height: '1080',
              width: '1920'
            },
          }
        ]
      }
    } );

    expect( wrapper.state( 'video' ).uploadStatus.error )
      .toEqual( true );
    expect( wrapper.find( 'li.video.error' ).exists() )
      .toEqual( true );
    expect( wrapper.contains(
      <Icon
        color="red"
        name="exclamation triangle"
        size="large"
      />
    ) ).toEqual( true );
    expect( wrapper.find( 'p' ).text() )
      .toEqual( 'An uploading error occurred for this item.' );
  } );

  it( 'renders error icon and message if `video` has an error', () => {
    const wrapper = shallow( Component );

    expect( wrapper.find( 'li.video.error' ).exists() )
      .toEqual( false );

    wrapper.setState( {
      video: {
        id: '3728',
        title: 'Made in America',
        uploadStatus: {
          error: false,
          success: false
        },
        error: true,
        source: [
          {
            burnedInCaptions: 'false',
            downloadUrl: 'https://video-download-url.com',
            duration: '9:16',
            filetype: '',
            md5: '',
            fileName: 'madeinamerica_english.mp4',
            size: {
              bitrate: 9832917,
              filesize: 662595174,
              height: '1080',
              width: '1920'
            },
          }
        ]
      }
    } );

    expect( wrapper.state( 'video' ).error ).toEqual( true );
    expect( wrapper.find( 'li.video.error' ).exists() )
      .toEqual( true );
    expect( wrapper.contains(
      <Icon
        color="red"
        name="exclamation triangle"
        size="large"
      />
    ) ).toEqual( true );
    expect( wrapper.find( 'p' ).text() )
      .toEqual( 'A loading error occurred for this item.' );
  } );
} );
