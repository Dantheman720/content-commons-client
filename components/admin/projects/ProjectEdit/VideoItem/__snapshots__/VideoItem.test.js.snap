// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoItem /> renders \`Progress\` component if \`isUploading\` 1`] = `
<li
  className="item video"
>
  <span
    className="modal-trigger"
    onClick={null}
    style={
      Object {
        "cursor": "not-allowed",
      }
    }
  >
    <div
      className="thumbnail isUploading"
    >
      <img
        alt="Man walking through factory (State Dept.)"
        src="https://staticcdp.s3.amazonaws.com/2018/05/courses.america.gov_1481/b3b38d194ff80d06dd837f57a41fe16f.jpg"
      />
      <p
        className="file-name"
      >
        madeinamerica_english_asdkjaf_al;kdflkeori_erpoiuzx,mnvz.mp4
      </p>
      <div
        className="loading-animation"
      >
        <Loader
          active={true}
          inline="centered"
        />
      </div>
    </div>
    <Progress
      active={true}
      color="blue"
      precision={0}
      progress={true}
      size="small"
      total={662595174}
      value={0}
    >
      <p>
        Upload in progress
      </p>
    </Progress>
    <h3
      className="item-heading ltr"
    >
      Made in America
    </h3>
    <p
      className="item-lang"
    >
      English
    </p>
  </span>
</li>
`;

exports[`<VideoItem /> renders a \`span\` with correct props if \`displayItemInModal\` is \`false\` 1`] = `
<li
  className="item video"
>
  <span
    className="wrapper"
    onClick={null}
    style={
      Object {
        "cursor": "default",
      }
    }
  >
    <div
      className="thumbnail"
    >
      <img
        alt="Man walking through factory (State Dept.)"
        src="https://staticcdp.s3.amazonaws.com/2018/05/courses.america.gov_1481/b3b38d194ff80d06dd837f57a41fe16f.jpg"
      />
      <p
        className="file-name"
      >
        madeinamerica_english_asdkjaf_al;kdflkeori_erpoiuzx,mnvz.mp4
      </p>
    </div>
    <h3
      className="item-heading ltr"
    >
      Made in America
    </h3>
    <p
      className="item-lang"
    >
      English
    </p>
  </span>
</li>
`;

exports[`<VideoItem /> renders without crashing 1`] = `
<li
  className="item video"
>
  <button
    className="modal-trigger"
    onClick={[MockFunction]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <div
      className="thumbnail"
    >
      <img
        alt="Man walking through factory (State Dept.)"
        src="https://staticcdp.s3.amazonaws.com/2018/05/courses.america.gov_1481/b3b38d194ff80d06dd837f57a41fe16f.jpg"
      />
      <p
        className="file-name"
      >
        madeinamerica_english_asdkjaf_al;kdflkeori_erpoiuzx,mnvz.mp4
      </p>
    </div>
    <h3
      className="item-heading ltr"
    >
      Made in America
    </h3>
    <p
      className="item-lang"
    >
      English
    </p>
  </button>
</li>
`;
